python/ray/_private/pydantic_compat.py

```diff
@@ -0,0 +1,107 @@
+# flake8: noqa
+import weakref
+from pkg_resources import packaging
+from typing import Any
+
+# Pydantic is a dependency of `ray["default"]` but not the minimal installation,
+# so handle the case where it isn't installed.
+try:
+    import pydantic
+
+    PYDANTIC_INSTALLED = True
+except ImportError:
+    pydantic = None
+    PYDANTIC_INSTALLED = False
+
+
+class CloudpickleableSchemaSerializer:
+    def __init__(self, schema, core_config):
+        self._schema = schema
+        self._core_config = core_config
+
+        from pydantic_core import SchemaSerializer
+
+        self._schema_serializer = SchemaSerializer(schema, core_config)
+
+    def __reduce__(self):
+        return CloudpickleableSchemaSerializer, (self._schema, self._core_config)
+
+    def __getattr__(self, attr: str):
+        return getattr(self._schema_serializer, attr)
+
+
+class WeakRefWrapper:
+    def __init__(self, obj: Any):
+        if obj is None:
+            self._wr = None
+        else:
+            self._wr = weakref.ref(obj)
+
+    def __reduce__(self):
+        return WeakRefWrapper, (self(),)
+
+    def __call__(self) -> Any:
+        if self._wr is None:
+            return None
+        else:
+            return self._wr()
+
+
+def monkeypatch_pydantic_2_for_cloudpickle():
+    """Patches non-serializable types introduced in Pydantic 2.0.
+
+    See https://github.com/pydantic/pydantic/issues/6763 for details.
+
+    This is a temporary workaround and will only work if Ray is imported *before*
+    Pydantic models are defined.
+    """
+    pydantic._internal._model_construction.SchemaSerializer = (
+        CloudpickleableSchemaSerializer
+    )
+    pydantic._internal._dataclasses.SchemaSerializer = CloudpickleableSchemaSerializer
+    pydantic.type_adapter.SchemaSerializer = CloudpickleableSchemaSerializer
+    pydantic._internal._model_construction._PydanticWeakRef = WeakRefWrapper
+
+
+if not PYDANTIC_INSTALLED:
+    IS_PYDANTIC_2 = False
```

Comment:
There is also some pydantic compat logic in `ray/util/state/common.py` that could be migrated to use the `IS_PYDANTIC_2` at least. Is that out of scope for this PR?

Reply:
Didn't see it, let me have a look

Reply:
I've migrated the logic in `ray/util/state/common.py`, thanks for the pointer